File: linkedlists/stack.py
from typing import Optional


class Node:
    def __init__(self, value, previous=None) -> None:
        self.value = value
        self.previous: Optional[Node] = previous


class Stack:
    def __init__(self) -> None:
        self.head: Optional[Node] = None
        self.length = 0

    def push(self, item):
        new_node = Node(item)

        if not self.head:
            self.head = new_node
            self.length += 1
            return

        new_node.previous = self.head
        self.head = new_node
        self.length += 1

    def pop(self):
        if not self.head:
            return None
        old_head = self.head
        self.head = self.head.previous
        self.length = max(0, self.length - 1)
        return old_head

    def peek(self):
        return self.head.value

File: linkedlists/stack.ts
type SNode<T> = {
  value: T;
  prev: SNode<T>;
};

export default class Stack<T> {
  public length: number;
  private head?: SNode<T>;

  constructor() {
    this.head = undefined;
    this.length = 0;
  }

  push(item: T): void {
    const new_node = { value: item } as SNode<T>;

    if (!this.head) {
      this.head = new_node;
      this.length++;
      return;
    }

    new_node.prev = this.head;
    this.head = new_node;
    this.length++;
  }

  pop(): T | undefined {
    this.length = Math.max(0, this.length - 1);
    if (!this.head) {
      return undefined;
    }
    const old_head = this.head;
    this.head = this.head.prev;
    return old_head.value;
  }

  peek(): T | undefined {
    return this.head?.value;
  }
}
